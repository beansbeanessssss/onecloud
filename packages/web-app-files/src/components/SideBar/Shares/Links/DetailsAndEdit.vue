<template>
  <div class="link-details oc-flex oc-flex-between oc-flex-middle">
    <div v-if="isModifiable">
      <link-role-dropdown
        :model-value="currentLinkRole"
        :available-role-options="availableRoleOptions"
        drop-offset="0"
        @update:model-value="updateSelectedRole"
      />
    </div>
    <p v-else class="oc-my-rm">
      <span
        v-oc-tooltip="$gettext(currentLinkRoleDescription)"
        class="link-current-role"
        v-text="$gettext(currentLinkRoleLabel)"
      />
    </p>
    <oc-checkbox
      v-if="isRunningOnEos && isCurrentLinkRoleUploader()"
      :model-value="currentLinkNotifyUploads"
      :label="$gettext('Notify me about uploads')"
      @input="toggleNotifyUploads"
    />
    <div :class="{ 'oc-pr-s': !isModifiable }" class="details-buttons">
      <oc-button
        v-if="link.indirect"
        v-oc-tooltip="viaTooltip"
        :area-label="viaTooltip"
        type="router-link"
        class="oc-files-file-link-via"
        appearance="raw"
        :to="viaRouterParams"
      >
        <oc-icon name="folder-shared" fill-type="line" />
      </oc-button>
      <oc-icon
        v-if="link.password"
        v-oc-tooltip="passwortProtectionTooltip"
        name="lock-password"
        fill-type="line"
        :aria-label="passwortProtectionTooltip"
      />
      <oc-icon
        v-if="link.expiration"
        v-oc-tooltip="expirationDateTooltip"
        class="oc-files-public-link-expires oc-ml-xs"
        :data-testid="`files-link-id-${link.id}-expiration-date`"
        :aria-label="expirationDateTooltip"
        name="calendar-event"
        fill-type="line"
      />
      <oc-icon
        v-if="isRunningOnEos && currentLinkNotifyUploadsExtraRecipients"
        v-oc-tooltip="$gettext('Uploads will be notified to a third party')"
        :aria-label="$gettext('Uploads will be notified to a third party')"
        name="mail-add"
        fill-type="line"
      />
      <div v-if="isModifiable">
        <oc-button
          :id="`edit-public-link-dropdown-toggl-${link.id}`"
          appearance="raw"
          class="edit-drop-trigger"
          :data-testid="`files-link-id-${link.id}-btn-edit`"
        >
          <oc-icon name="more-2" />
        </oc-button>
        <oc-drop
          ref="editPublicLinkDropdown"
          :drop-id="`edit-public-link-dropdown`"
          :toggle="`#edit-public-link-dropdown-toggl-${link.id}`"
          padding-size="small"
          mode="click"
        >
          <oc-list
            v-if="editOptions.length > 0"
            class="edit-public-link-dropdown-menu oc-pb-s oc-files-context-actions-border"
          >
            <li
              v-for="(option, i) in editOptions"
              :key="`public-link-edit-option-${i}`"
              class="oc-rounded oc-menu-item-hover"
            >
              <oc-datepicker
                v-if="option.showDatepicker"
                v-model="newExpiration"
                class="link-expiry-picker oc-flex oc-width-1-1"
                :min-date="expirationRules.min"
                :max-date="expirationRules.max"
                :locale="$language.current"
                :is-required="expirationRules.enforced"
              >
                <template #default="{ togglePopover }">
                  <oc-button
                    :data-testid="`files-link-id-${link.id}-edit-${option.id}`"
                    appearance="raw"
                    class="oc-p-s action-menu-item"
                    :variation="option.variation"
                    @click="togglePopover"
                  >
                    <oc-icon :name="option.icon" fill-type="line" size="medium" />
                    <span v-text="option.title" />
                  </oc-button>
                  <oc-button
                    v-if="option.remove && option.remove.isRemovable"
                    :data-testid="`files-link-id-${link.id}-edit-${option.id}`"
                    :aria-label="option.remove.title"
                    appearance="raw"
                    @click="option.remove.method"
                  >
                    <oc-icon :name="option.remove.icon" />
                  </oc-button>
                </template>
              </oc-datepicker>
              <oc-button
                v-else
                appearance="raw"
                class="oc-p-s action-menu-item"
                :data-testid="`files-link-id-${link.id}-edit-${option.id}`"
                @click="option.method"
              >
                <oc-icon :name="option.icon" fill-type="line" size="medium" />
                <span v-text="option.title" />
              </oc-button>
            </li>
          </oc-list>
          <oc-list
            class="edit-public-link-dropdown-menu"
            :class="{ 'oc-pt-s': editOptions.length > 0 }"
          >
            <li class="oc-rounded oc-menu-item-hover">
              <oc-button
                appearance="raw"
                class="oc-p-s action-menu-item"
                :data-testid="`files-link-id-${link.id}-edit-${deleteOption.id}`"
                @click="deleteOption.method"
              >
                <oc-icon :name="deleteOption.icon" fill-type="line" size="medium" />
                <span v-text="deleteOption.title" />
              </oc-button>
            </li>
          </oc-list>
        </oc-drop>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { basename } from 'path'
import { DateTime } from 'luxon'
import { mapGetters } from 'vuex'
import * as EmailValidator from 'email-validator'
import {
  createLocationSpaces,
  useConfigurationManager,
  LinkRoleDropdown,
  useModals
} from '@ownclouders/web-pkg'
import {
  linkRoleInternalFile,
  linkRoleInternalFolder,
  linkRoleUploaderFolder,
  LinkShareRoles,
  ShareRole
} from '@ownclouders/web-client/src/helpers/share'
import { computed, defineComponent, inject, PropType, Ref, ref } from 'vue'
import { formatDateFromDateTime, formatRelativeDateFromDateTime } from '@ownclouders/web-pkg'
import { Resource, SpaceResource } from '@ownclouders/web-client/src/helpers'
import { createFileRouteOptions } from '@ownclouders/web-pkg'
import { OcDrop } from 'design-system/src/components'
import { usePasswordPolicyService, ExpirationRules } from '@ownclouders/web-pkg'
import { useGettext } from 'vue3-gettext'
import SetLinkPasswordModal from '../../../Modals/SetLinkPasswordModal.vue'

export default defineComponent({
  name: 'DetailsAndEdit',
  components: { LinkRoleDropdown },
  props: {
    availableRoleOptions: {
      type: Array as PropType<ShareRole[]>,
      required: true
    },
    canRename: {
      type: Boolean,
      default: false
    },
    expirationRules: {
      type: Object as PropType<ExpirationRules>,
      required: true
    },
    isFolderShare: {
      type: Boolean,
      default: false
    },
    isModifiable: {
      type: Boolean,
      default: false
    },
    isPasswordEnforced: {
      type: Boolean,
      default: false
    },
    isPasswordRemovable: {
      type: Boolean,
      default: false
    },
    link: {
      type: Object,
      required: true
    }
  },
  emits: ['removePublicLink', 'updateLink'],
  setup(props, { emit }) {
    const { registerModal } = useModals()
    const { $gettext, current } = useGettext()
    const configurationManager = useConfigurationManager()
    const passwordPolicyService = usePasswordPolicyService()

    const currentLinkRole = ref<ShareRole>(
      LinkShareRoles.getByBitmask(props.link.permissions, props.isFolderShare)
    ) as Ref<ShareRole>

    const dateExpire = computed(() => {
      return formatRelativeDateFromDateTime(
        DateTime.fromISO(props.link.expiration).endOf('day'),
        current
      )
    })

    const updateLink = ({ link, onSuccess = () => {} }) => {
      link = link || props.link
      emit('updateLink', { link, onSuccess })
    }
    const updateSelectedRole = (role: ShareRole) => {
      currentLinkRole.value = role
      updateLink({
        link: { ...props.link, permissions: role.bitmask(false) }
      })
    }

    const showPasswordModal = () => {
      registerModal({
        title: props.link.password ? $gettext('Edit password') : $gettext('Add password'),
        customComponent: SetLinkPasswordModal,
        customComponentAttrs: () => ({ link: props.link })
      })
    }

    return {
      space: inject<Ref<SpaceResource>>('space'),
      resource: inject<Ref<Resource>>('resource'),
      passwordPolicyService,
      dateExpire,
      updateLink,
      updateSelectedRole,
      currentLinkRole,
      isRunningOnEos: computed(() => configurationManager.options.isRunningOnEos),
      showPasswordModal,
      registerModal
    }
  },
  data() {
    return {
      newExpiration: this.link.expiration
    }
  },
  computed: {
    ...mapGetters('runtime/spaces', ['spaces']),

    currentLinkRoleDescription() {
      return this.currentLinkRole?.description(false) || ''
    },

    currentLinkRoleLabel() {
      if (this.currentLinkRole?.longLabel !== '') {
        return this.$gettext(this.currentLinkRole.longLabel)
      }
      return this.$gettext(this.currentLinkRole?.label || '')
    },

    editOptions() {
      const result = []

      if (this.canRename) {
        result.push({
          id: 'rename',
          title: this.$gettext('Rename'),
          icon: 'pencil',
          method: this.showRenameModal
        })
      }

      if (this.link.expiration) {
        result.push({
          id: 'edit-expiration',
          title: this.$gettext('Expires %{expires}', { expires: this.dateExpire }),
          method: this.updateLink,
          icon: 'calendar-event',
          showDatepicker: true,
          remove: {
            id: 'remove-expiration',
            title: this.$gettext('Remove expiration date'),
            icon: 'close',
            isRemovable: !this.expirationRules.enforced,
            method: () =>
              this.updateLink({
                link: {
                  ...this.link,
                  expiration: ''
                }
              })
          }
        })
      } else if (!this.isAliasLink) {
        result.push({
          id: 'add-expiration',
          title: this.$gettext('Set expiration date'),
          method: this.updateLink,
          icon: 'calendar-event',
          showDatepicker: true
        })
      }

      if (this.link.password) {
        result.push({
          id: 'edit-password',
          title: this.$gettext('Edit password'),
          icon: 'lock-password',
          method: this.showPasswordModal
        })

        if (this.isPasswordRemovable) {
          result.push({
            id: 'remove-password',
            title: this.$gettext('Remove password'),
            icon: 'lock-unlock',
            method: () =>
              this.updateLink({
                link: {
                  ...this.link,
                  password: ''
                }
              })
          })
        }
      }
      if (!this.link.password && !this.isAliasLink) {
        result.push({
          id: 'add-password',
          title: this.$gettext('Add password'),
          icon: 'lock-password',
          method: this.showPasswordModal
        })
      }

      if (this.isCurrentLinkRoleUploader && this.currentLinkNotifyUploads) {
        result.push({
          id: 'add-notify-uploads-extra-recipients',
          title: this.notifyUploadsExtraRecipientsMenuEntry,
          icon: 'mail-add',
          method: this.showNotifyUploadsExtraRecipientsModal
        })
      }
      if (this.currentLinkNotifyUploadsExtraRecipients) {
        result.push({
          id: 'remove-notify-uploads-extra-recipients',
          title: this.$gettext('Remove third party notification'),
          icon: 'mail-close',
          method: () =>
            this.updateLink({
              link: {
                ...this.link,
                notifyUploadsExtraRecipients: ''
              }
            })
        })
      }

      return result
    },

    deleteOption() {
      return {
        id: 'delete',
        title: this.$gettext('Delete link'),
        method: this.deleteLink,
        icon: 'delete-bin-5',
        variation: 'danger'
      }
    },

    viaRouterParams() {
      const matchingSpace = (this.space ||
        this.spaces.find((space) => space.id === this.resource.storageId)) as SpaceResource
      if (!matchingSpace) {
        return {}
      }

      return createLocationSpaces(
        'files-spaces-generic',
        createFileRouteOptions(matchingSpace, {
          path: this.link.path,
          fileId: this.link.file.source
        })
      )
    },

    localExpirationDate() {
      return formatDateFromDateTime(
        DateTime.fromISO(this.link.expiration).endOf('day'),
        this.$language.current
      )
    },

    expirationDateRelative() {
      return formatRelativeDateFromDateTime(
        DateTime.fromISO(this.link.expiration).endOf('day'),
        this.$language.current
      )
    },

    expirationDateTooltip() {
      return this.$gettext(
        'Expires %{timeToExpiry} (%{expiryDate})',
        { timeToExpiry: this.expirationDateRelative, expiryDate: this.localExpirationDate },
        true
      )
    },

    viaTooltip() {
      if (!this.link.indirect) {
        return null
      }
      return (
        this.$gettext('Navigate to the parent (%{folderName})'),
        { folderName: basename(this.link.path) },
        true
      )
    },

    passwortProtectionTooltip() {
      return this.$gettext('This link is password-protected')
    },

    isAliasLink() {
      if (this.isFolderShare) {
        return parseInt(this.link.permissions) == linkRoleInternalFolder.bitmask(false)
      }
      return parseInt(this.link.permissions) == linkRoleInternalFile.bitmask(false)
    },

    currentLinkNotifyUploads() {
      return this.link.notifyUploads
    },
    currentLinkNotifyUploadsExtraRecipients() {
      return this.link.notifyUploadsExtraRecipients
    },
    notifyUploadsExtraRecipientsMenuEntry() {
      if (this.currentLinkNotifyUploadsExtraRecipients) {
        return this.$gettext('Edit third party notification')
      }
      return this.$gettext('Notify a third party about uploads')
    }
  },
  watch: {
    newExpiration(expiration) {
      this.updateLink({
        link: {
          ...this.link,
          expiration
        }
      })
    }
  },
  methods: {
    deleteLink() {
      this.$emit('removePublicLink', { link: this.link })
      ;(this.$refs.editPublicLinkDropdown as InstanceType<typeof OcDrop>).hide()
    },
    showRenameModal() {
      this.registerModal({
        title: this.$gettext('Edit name'),
        confirmText: this.$gettext('Save'),
        hasInput: true,
        inputValue: this.link.name,
        inputLabel: this.$gettext('Link name'),
        onInput: (name, setError) => {
          if (name.length > 255) {
            return setError(this.$gettext('Link name cannot exceed 255 characters'))
          }
          return setError(null)
        },
        onConfirm: (name: string) =>
          Promise.resolve(
            this.updateLink({
              link: {
                ...this.link,
                name
              }
            })
          )
      })
    },

    toggleNotifyUploads() {
      if (this.currentLinkNotifyUploads) {
        this.$emit('updateLink', {
          link: { ...this.link, notifyUploads: false, notifyUploadsExtraRecipients: '' }
        })
      } else {
        this.$emit('updateLink', { link: { ...this.link, notifyUploads: true } })
      }
    },
    isCurrentLinkRoleUploader() {
      return (
        LinkShareRoles.getByBitmask(parseInt(this.link.permissions), this.isFolderShare).bitmask(
          false
        ) === linkRoleUploaderFolder.bitmask(false)
      )
    },
    showNotifyUploadsExtraRecipientsModal() {
      this.registerModal({
        icon: 'mail-add',
        title: this.$gettext('Notify a third party about uploads'),
        confirmText: this.$gettext('Apply'),
        hasInput: true,
        inputDescription: this.$gettext(
          'When a file is uploaded, this address will be notified as well.'
        ),
        inputValue: this.currentLinkNotifyUploadsExtraRecipients,
        inputLabel: this.$gettext('Email address'),
        inputType: 'email',
        onInput: (value, setError) => setError(this.getEmailValidationMsg(value)),
        onConfirm: (value: string) =>
          Promise.resolve(
            this.updateLink({
              link: {
                ...this.link,
                notifyUploadsExtraRecipients: value
              }
            })
          )
      })
    },
    getEmailValidationMsg(email: string) {
      if (!EmailValidator.validate(email)) {
        return this.$gettext('Email is invalid')
      }
      if (email === '') {
        return this.$gettext("Email can't be empty")
      }
      return null
    }
  }
})
</script>

<style lang="scss" scoped>
.details-buttons {
  min-width: 5rem !important;
  display: flex;
  justify-content: flex-end;
}

.edit-public-link-dropdown-menu {
  .action-menu-item {
    width: 100%;
    justify-content: flex-start;
  }
}

.oc-files-public-link-expires {
  margin-top: 1px; // to align with the other elements
}
</style>
