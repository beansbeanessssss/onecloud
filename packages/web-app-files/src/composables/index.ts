export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './sort'
export * from './viewMode'
