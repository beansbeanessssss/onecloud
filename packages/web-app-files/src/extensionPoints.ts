import { ExtensionPoint } from '@ownclouders/web-pkg'
import { computed } from 'vue'

export const uploadMenuExtensionPoint: ExtensionPoint = {
  id: 'app.files.upload-menu',
  extensionType: 'action',
  multiple: true
}
export const quickActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.quick-actions',
  extensionType: 'action',
  multiple: true
}
export const batchActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.batch-actions',
  extensionType: 'action',
  multiple: true
}
export const contextActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.context-actions',
  extensionType: 'action',
  multiple: true
}
export const fileSideBarExtensionPoint: ExtensionPoint = {
  id: 'app.files.sidebar',
  extensionType: 'sidebarPanel',
  multiple: true
}
export const folderViewsFolderExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.folder',
  extensionType: 'folderView'
}
export const folderViewsFavoritesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.favorites',
  extensionType: 'folderView'
}
export const folderViewsProjectSpacesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.project-spaces',
  extensionType: 'folderView'
}

export const extensionPoints = () => {
  return computed<ExtensionPoint[]>(() => {
    return [
      uploadMenuExtensionPoint,
      quickActionsExtensionPoint,
      batchActionsExtensionPoint,
      contextActionsExtensionPoint,
      fileSideBarExtensionPoint,
      folderViewsFolderExtensionPoint,
      folderViewsFavoritesExtensionPoint,
      folderViewsProjectSpacesExtensionPoint
    ]
  })
}
