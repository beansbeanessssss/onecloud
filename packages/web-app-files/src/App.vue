<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from '@vue/composition-api'
import { Route } from 'vue-router'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { bus } from 'web-pkg/src/instance'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), (to: Route, from?: Route) => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = bus.subscribe('drag-over', this.onDragOver)
    const dragOut = bus.subscribe('drag-out', this.hideDropzone)
    const drop = bus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('drag-over', dragOver)
      bus.unsubscribe('drag-out', dragOut)
      bus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
