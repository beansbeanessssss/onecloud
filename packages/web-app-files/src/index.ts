import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import { defineWebApplication, quickActions } from '@ownclouders/web-pkg'
import store from './store'
import { extensions } from './extensions'
import fileSideBars from './fileSideBars'
import { buildRoutes } from '@ownclouders/web-pkg'
import { AppNavigationItem } from '@ownclouders/web-pkg'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import {
  SpaceResource,
  isPersonalSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  color: 'var(--oc-color-swatch-primary-muted)',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
export const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      isActive: () => {
        return (
          !context.$store.getters['runtime/spaces/currentSpace'] ||
          context.$store.getters['runtime/spaces/currentSpace']?.isOwner(
            context.$store.getters.user
          )
        )
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find(
          (drive) => isPersonalSpaceResource(drive) && drive.isOwner(context.$store.getters.user)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      isActive: () => {
        const space = context.$store.getters['runtime/spaces/currentSpace'] as SpaceResource
        // last check is when fullShareOwnerPaths is enabled
        return !space || isShareSpaceResource(space) || !space?.isOwner(context.$store.getters.user)
      },
      activeFor: [
        { path: `/${appInfo.id}/spaces/share` },
        { path: `/${appInfo.id}/spaces/personal` }
      ],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0' && capabilities.files?.undelete
      }
    }
  ]
}

export default defineWebApplication({
  setup(args) {
    return {
      appInfo,
      store,
      routes: buildRoutes({
        App,
        Favorites,
        FilesDrop,
        SearchResults,
        Shares: {
          SharedViaLink,
          SharedWithMe,
          SharedWithOthers
        },
        Spaces: {
          DriveResolver: SpaceDriveResolver,
          Projects: SpaceProjects
        },
        Trash: {
          Overview: TrashOverview
        }
      }),
      navItems,
      quickActions,
      translations,
      extensions: extensions(args)
    }
  }
})
