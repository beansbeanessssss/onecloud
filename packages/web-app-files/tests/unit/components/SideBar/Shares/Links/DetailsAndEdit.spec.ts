import DetailsAndEdit from 'web-app-files/src/components/SideBar/Shares/Links/DetailsAndEdit.vue'
import { LinkShare, ShareRoleNG } from '@ownclouders/web-client/src/helpers/share'
import { defaultPlugins, shallowMount, defaultComponentMocks } from 'web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { useLinkTypes } from '@ownclouders/web-pkg'
import { SharingLinkType } from '@ownclouders/web-client/src/generated'

vi.mock('@ownclouders/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useLinkTypes: vi.fn()
}))

const exampleLink = {
  displayName: 'Example link',
  webUrl: 'https://some-url.com/abc',
  type: SharingLinkType.View
} as LinkShare

describe('DetailsAndEdit component', () => {
  describe('if user can not edit', () => {
    it('does not render dropdown or edit button', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink)
      expect(wrapper.find('link-role-dropdown-stub').exists()).toBeFalsy()
      expect(wrapper.find('.edit-drop-trigger').exists()).toBeFalsy()
    })
  })

  describe('if user can edit', () => {
    it('renders dropdown and edit button', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      expect(wrapper.find('link-role-dropdown-stub').exists()).toBeTruthy()
      expect(wrapper.find('.edit-drop-trigger').exists()).toBeTruthy()
    })

    it.todo('test edit options, button clicks and event handling/propagation')
  })

  describe('editOptions computed property', () => {
    it('does not add "add-expiration" option if isAliasLink is true', () => {
      const exampleLinkInternal = { ...exampleLink }
      exampleLinkInternal.type = SharingLinkType.Internal
      const { wrapper } = getShallowMountedWrapper(exampleLinkInternal, false, true)
      expect(wrapper.vm.editOptions.some((option) => option.id === 'add-expiration')).toBe(false)
    })

    it('adds "add-expiration" option if isAliasLink is false', () => {
      const { wrapper } = getShallowMountedWrapper(exampleLink, false, true)
      expect(wrapper.vm.editOptions.some((option) => option.id === 'add-expiration')).toBe(true)
    })
  })
})

function getShallowMountedWrapper(
  linkShare: LinkShare,
  expireDateEnforced = false,
  isModifiable = false
) {
  vi.mocked(useLinkTypes).mockReturnValue(
    mock<ReturnType<typeof useLinkTypes>>({
      getAvailableLinkTypes: () => [SharingLinkType.View],
      getLinkRoleByType: () => mock<ShareRoleNG>({ displayName: '', description: '', label: '' })
    })
  )

  const mocks = defaultComponentMocks()
  return {
    wrapper: shallowMount(DetailsAndEdit, {
      props: {
        canRename: true,
        expirationRules: {
          enforced: expireDateEnforced,
          default: null,
          min: 'Wed Apr 01 2020 00:00:00 GMT+0000 (Coordinated Universal Time)',
          max: null
        },
        linkShare,
        isModifiable,
        isPasswordEnforced: false
      },
      global: {
        mocks,
        renderStubDefaultSlot: true,
        stubs: { OcDatepicker: false, 'date-picker': true },
        plugins: [...defaultPlugins()],
        provide: mocks
      }
    })
  }
}
