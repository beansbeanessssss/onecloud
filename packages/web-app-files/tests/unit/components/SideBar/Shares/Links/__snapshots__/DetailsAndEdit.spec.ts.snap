// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsAndEdit component if user can edit renders dropdown and edit button 1`] = `
<div class="link-details oc-flex oc-flex-between oc-flex-middle oc-pl-s">
  <div>
    <oc-button-stub appearance="raw" class="edit-public-link-role-dropdown-toggle oc-text-left" gapsize="none" id="edit-public-link-role-dropdown-toggle-undefined" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
      <span class="link-current-role">Viewer</span>
      <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="arrow-down-s" size="medium" type="span" variation="passive"></oc-icon-stub>
    </oc-button-stub>
    <oc-drop-stub class="edit-public-link-role-dropdown" dropid="edit-public-link-role-dropdown" mode="click" offset="0" paddingsize="small" position="bottom-start" toggle="#edit-public-link-role-dropdown-toggle-undefined">
      <oc-list-stub class="roleDropdownList">
        <li>
          <oc-button-stub appearance="raw" class="oc-p-s" gapsize="medium" id="files-role-internal" justifycontent="space-between" size="medium" submit="button" type="button" variation="passive">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="user" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Internal</span>
                <span class="oc-text-small">People need to be invited and login is required</span>
              </span>
            </span>
            <span class="oc-flex">
              <!--v-if-->
            </span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw" class="selected oc-background-primary-gradient oc-p-s" gapsize="medium" id="files-role-viewer" justifycontent="space-between" size="medium" submit="button" type="button" variation="inverse">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="eye" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Viewer</span>
                <span class="oc-text-small">Recipients can view and download contents.</span>
              </span>
            </span>
            <span class="oc-flex">
              <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="check" size="medium" type="span" variation="passive"></oc-icon-stub>
            </span>
          </oc-button-stub>
        </li>
        <li>
          <oc-button-stub appearance="raw" class="oc-p-s" gapsize="medium" id="files-role-editor" justifycontent="space-between" size="medium" submit="button" type="button" variation="passive">
            <span class="oc-flex oc-flex-middle">
              <oc-icon-stub accessiblelabel="" class="oc-pl-s oc-pr-m" color="" filltype="fill" name="pencil" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>
                <span class="oc-text-bold oc-display-block oc-width-1-1">Editor</span>
                <span class="oc-text-small">Recipients can view, download and edit contents.</span>
              </span>
            </span>
            <span class="oc-flex">
              <!--v-if-->
            </span>
          </oc-button-stub>
        </li>
      </oc-list-stub>
    </oc-drop-stub>
  </div>
  <div class="details-buttons">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <div>
      <oc-button-stub appearance="raw" class="edit-drop-trigger" gapsize="medium" id="edit-public-link-dropdown-toggl-undefined" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
        <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="more-2" size="medium" type="span" variation="passive"></oc-icon-stub>
      </oc-button-stub>
      <oc-drop-stub dropid="edit-public-link-dropdown" mode="click" paddingsize="small" position="bottom-start" toggle="#edit-public-link-dropdown-toggl-undefined">
        <oc-list-stub class="edit-public-link-dropdown-menu oc-files-context-actions-border oc-pb-s">
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="pencil" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Rename</span>
            </oc-button-stub>
          </li>
          <li class="oc-rounded oc-menu-item-hover">
            <date-picker-stub class="link-expiry-picker oc-datepicker link-expiry-picker" inputdebounce="1000" locale="en" min-date="Wed Apr 01 2020 00:00:00 GMT+0000 (Coordinated Universal Time)" mode="date" modelconfig="[object Object]" popover="[object Object]" updateoninput="true">
              <!-- @slot Default slot to use as the popover anchor for datepicker -->
              <!-- args is undefined during initial render, hence we check it here -->
              <!--v-if-->
            </date-picker-stub>
          </li>
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="lock-password" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Add password</span>
            </oc-button-stub>
          </li>
        </oc-list-stub>
        <oc-list-stub class="edit-public-link-dropdown-menu oc-pt-s">
          <li class="oc-rounded oc-menu-item-hover">
            <oc-button-stub appearance="raw" class="oc-p-s action-menu-item" gapsize="medium" justifycontent="center" size="medium" submit="button" type="button" variation="passive">
              <oc-icon-stub accessiblelabel="" color="" filltype="line" name="delete-bin-5" size="medium" type="span" variation="passive"></oc-icon-stub>
              <span>Delete link</span>
            </oc-button-stub>
          </li>
        </oc-list-stub>
      </oc-drop-stub>
    </div>
  </div>
</div>
`;

exports[`DetailsAndEdit component if user can not edit does not render dropdown or edit button 1`] = `
<div class="link-details oc-flex oc-flex-between oc-flex-middle oc-pl-s">
  <p class="oc-my-rm">
    <span class="link-current-role">Viewer</span>
  </p>
  <div class="oc-pr-s details-buttons">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
